pallets/dlc-machine/src/lib.rs

```diff
@@ -11,43 +11,190 @@ mod tests;
+    #[pallet::getter(fn machine_id_2_gpu_count_of_idc_machine_nft_staking)]
+    pub type MachineId2GPUCountInStakingOfIDCMachineNFTStaking<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        PhaseLevel,
+        Blake2_128Concat,
+        MachineId,
+        u32,
+        ValueQuery,
+    >;
 
     #[pallet::pallet]
     #[pallet::without_storage_info]
     pub struct Pallet<T>(_);
 
     #[pallet::event]
-    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
     pub enum Event<T: Config> {
         ReportDLCStaking(T::AccountId, MachineId),
     }
 
+    #[pallet::error]
+    pub enum Error<T> {
+        InvalidValue,
+        RenterNotOwner,
+        AlreadyInStaking,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        fn on_initialize(block_number: T::BlockNumber) -> Weight {
+            if block_number.saturated_into::<u64>() % ONE_HOUR as u64 == 0 {
+                frame_support::log::info!("🔍 start clean up machine_ids_in_staking");
+                let machine_id_to_remove: Vec<MachineId> = DLCMachinesInStaking::<T>::iter_keys()
+                    .filter(|machine_id| {
+                        let machine_info_result =
+                            online_profile::Pallet::<T>::machines_info(machine_id);
+                        if let Some(machine_info) = machine_info_result {
+                            // must be rented
+                            if machine_info.machine_status != MachineStatus::Rented {
+                                return true
+                            };
+
+                            // must be rented by owner
+                            if machine_info.renters.len() != 1 {
+                                return true
+                            }
+
+                            let renter_result = machine_info.renters.last();
+                            if renter_result.is_none() {
+                                return true
+                            }
+
+                            let renter = renter_result.unwrap().clone();
+                            if machine_info.controller != renter &&
+                                machine_info.machine_stash != renter
+                            {
+                                return true
+                            }
+                        } else {
+                            return true
+                        }
+
+                        return false
+                    })
+                    .collect();
+
+                for machine_id in machine_id_to_remove {
+                    frame_support::log::info!(
+                        "remove machine_id : {}",
+                        str::from_utf8(&machine_id).unwrap()
+                    );
+                    DLCMachinesOwnerRentEnded::<T>::insert(machine_id, ());
+
+                    // DLCMachinesInStaking::<T>::remove(machine_id);
```

Comment:
should emit event here.
```suggestion
                    DLCMachinesInStaking::<T>::remove(machine_id);
                    Self::deposit_event(Event:: DLCMachinesOwnerRentEnded(machine_id));
```